test/unit/org/apache/cassandra/index/sai/cql/StorageAttachedIndexDDLTest.java

```diff
@@ -1311,6 +1312,40 @@ public void nodetoolStopInitialIndexBuild() throws Throwable
         assertTrue(verifyChecksum(numericIndexTermType, numericIndexIdentifier));
     }
 
+    @Test
+    public void shouldMarkQueryableInInitializationTask() throws Throwable
+    {
+        createTable(CREATE_TABLE_TEMPLATE);
+        disableCompaction(KEYSPACE);
+        IndexIdentifier idxIdentifier = createIndexIdentifier(createIndexAsync(String.format(CREATE_INDEX_TEMPLATE, "v1")));
+
+        // create 10 SSTables
+        for (int i = 0; i < 10; i++)
+        {
+            execute("INSERT INTO %s (id1, v1, v2) VALUES (?, ?, ?)", String.valueOf(i), i, String.valueOf(i));
+            flush();
+        }
+
+        ResultSet rows = executeNet("SELECT id1 FROM %s WHERE v1 >= 5");
+        assertEquals(5, rows.all().size());
+
+        // Make the index artificially non-queryable:
+        ColumnFamilyStore cfs = getCurrentColumnFamilyStore();
+        Index index = cfs.indexManager.getIndexByName(idxIdentifier.indexName);
+        cfs.indexManager.makeIndexNonQueryable(index, Index.Status.BUILD_FAILED);
+
+        // Query should fail with the index in an artificially non-queryable state:
+        assertThatThrownBy(() -> executeNet("SELECT id1 FROM %s WHERE v1 >= 5")).isInstanceOf(ReadFailureException.class);
+
+        // Node must be in STARTING mode for it to be necessary for the initialization task to pre-emptively validate:
+        StorageService.instance.setStartingModeUnsafe();
+        // Simply getting the initialization task (and not running it) will validate and mark the index queryable again:
+        cfs.indexManager.buildIndex(index);
```

Comment:
Note here that `buildIndex()` will have started the asynchronous index build task by the time this returns, but the test should always fail without the patch, given the boilerplate there before it actually starts validating the exiting index. It's *possible* the pre-patch code could have passed the test if we got really lucky and paused for a significant amount of time right here after returning from `buildIndex()`.